### PR TITLE
Add get/set API to the Context and make it coroutine-safe

### LINKED ISSUE #15276: [Feature Request]: Better API for `Context`
### Feature Description

Implement a get/set API for the `Context` instance instead of accessing the underlying `data` dictionary directly, e.g. `context.get("foo")` instead of `context.data["foo"]`

### Reason

Using a vanilla dict for context data is limiting and not very intuitive. This change would also allow for more flexible data persistence strategies, as the context data wouldn't be confined to in-memory storage in the current process.

### Value of Feature

_No response_

### PR DESCRIPTION
**Pull request by masci**
# Description

- Introduce `get` and `set` APIs and make them coroutine-safe
- Hide the actual storage data structures, so that the context can be implemented differently without breaking changes
- Introduce the concept of "private storaging" for data that shouldn't be shared across different steps

Fixes #15276

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-core/llama_index/core/workflow/context.py [hunk 1]
```diff
@@ -1,20 +1,87 @@
...
-        self.parent = parent
+        # Local data storage
+        self._locals: Dict[str, Any] = {}
+
+        # Local configuration
+        self._parent: Optional[Context] = parent
         self._events_buffer: Dict[Type[Event], Event] = {}
 
+    @property
+    def parent(self) -> Optional["Context"]:
```
Comment (logan-markewich) [inline llama-index-core/llama_index/core/workflow/context.py:42]: We might not want to allow parent context access

If we think of nesting workflows, coupling them too much will make testing harder for the developer no?

instead, I would encourage the parent workflow passing in whatever is needed to run the child workflow 

(I might be misunderstanding the intention of this too though)

Comment (masci) [reply llama-index-core/llama_index/core/workflow/context.py:42]: This is just to access the lock, I can make this property `def lock(self)` and avoid exposing the parent context 👍

Comment (logan-markewich) [reply llama-index-core/llama_index/core/workflow/context.py:42]: yea that might be a bit more straightforward 👀

File: llama-index-core/llama_index/core/workflow/context.py [hunk 2]
```diff
@@ -1,21 +1,94 @@
...
+            key: A unique string to identify the value stored.
+            default: The value to return when `key` is missing instead of raising an exception.
+
+        Raises:
+            ValueError: When there's not value accessible corresponding to `key`.
+        """
+        if key in self._locals:
+            return self._locals[key]
+        elif key in self._globals:
+            await self.lock.acquire()
```
Comment (logan-markewich) [inline llama-index-core/llama_index/core/workflow/context.py:69]: It might be slightly safer to do `with self.lock` instead?

**Review by logan-markewich — APPROVED** [review]
This makes sense to me 👍🏻

**Review by logan-markewich — APPROVED** [review]

**Pull request merged**